app/javascript/application.js

```diff
@@ -11,6 +11,7 @@
 import 'jquery';
 import 'jquery-ujs'
 import 'select2';
+import 'select2/dist/js/i18n/de'
```

Comment:
Do you think it is possible and smart to use the available (and imported) I18n library below to load that dynamically?

Reply:
I don't know. I keep getting weird errors when trying to. I tried the following:
```
import { I18n } from "i18n-js";
import locales from "../../tmp/locales.json";

locales.forEach(locale => import(`select2/dist/js/i18n/${locale}`));
```
also
```
const available_locales = ['en', 'de'];
for(let i = 0; i < available_locales.length; i++) {
  import(`select2/dist/js/i18n/${available_locales[i]}`);
}
```
and even this:
```
const locale = 'de'
import('select2/dist/js/i18n/' + locale);
```
however this works:
```
const locale = 'de'
import('select2/dist/js/i18n/' + 'de');
```

The error I am getting is:
```
ERROR in ./node_modules/select2/dist/js/i18n/build.txt 3:14
Module parse failed: Assigning to rvalue (3:14)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| 
| af.js
> ----------------
| af.js
| 

webpack 5.90.2 compiled with 1 error in 408 ms
```

So I think I am misunderstanding something about dynamic imports or about how webpack handles them.

Reply:
Uh, that's a though one. I think, you didn't get that wrong. The dynamic import is, per se, working correctly as far as I can tell. However, it's too "smart" and including too many files 😕

The issue here is the select2 package and how webpack prepares the file(s) for dynamic inclusion. In short, webpack prepares each language file separately as a chunk, so that it can be loaded (if needed) by the browser. That's desired, since it prevents browsers from loading (network-wise) and interpreting (locally) the files not required. Still, some preprocessing is required for these files to work as expected.

Now, webpack encounters issues when accessing the `/node_modules/select2/dist/js/i18n/build.txt` file. That's because the file is processed as a JavaScript file (it might be, potentially, included through our `application.js`), but you'll likely agree that this doesn't make sense (if you open the file, you'll just find a list of those file names, such as `en.js`, `de.js`, ...). If you ask me, the presence of the file is a build error of the upstream repo, and potentially can be reported there as well. Nonetheless, we need to tell webpack to handle that file differently.

I tried various things, but the "best" solution I came up with so far is to modify the webpack config. To do so, place the following code snippet in `config/webpack/webpack.config.js`, lines 26+

```js
        rules: [
            erb,
            {
                test: /select2\/dist\/js\/i18n\/build\.txt$/,
                type: 'asset/resource',
                generator: {
                    emit: false,
                },
            },
        ]
```

Of course, this can also be extracted to a dedicated "loader", similar to how we handle the "erb" thing. With this rule, we will add a dedicated handler for the `build.txt` file, and prevent it from being emitted altogether. 

In summary, the aforementioned config change will allow you to use the dynamic loading as expected. Since I am not a true webpack expert, that's all I can offer (otherwise, you'd need to ask in #dev-help).

<hr>

While testing this integration, I further noticed some race condition (at least locally): Sometimes, the translation for German was not working as expected. I suspect some race condition, based on loading the file (with the translation) and firing the `turbolinks:ready` event, which will initialise the Select2 element. Since I was testing with tasks#new, I modified `app/assets/javascripts/labels.coffee.erb` to wait another second:

```coffee
# Previous code
# $(document).on 'turbolinks:load', ready

# Modified code for testing:
$(document).on 'turbolinks:load', () ->
  setTimeout(ready, 1000);
```

Reply:
The dynamic import now works but I cannot recreate the race condition. What exactly is happening? Is the translation English even though it should be German?

Reply:
Yes, the translation is still in English besides the remaining page being in German. There is no error in the console on the language is correctly loaded:

<img width="1840" alt="Bildschirmfoto 2024-03-09 um 18 03 16" src="https://github.com/openHPI/codeharbor/assets/7300329/d4421dc9-0b62-40df-a347-ee62301a17e9">

It just required me five tries to get the desired behavior (German locale with German select2). For this, I was "just" reloading the page multiple times and trying over and over again:

<img width="1840" alt="Bildschirmfoto 2024-03-09 um 18 04 53" src="https://github.com/openHPI/codeharbor/assets/7300329/26b770f7-06e6-40a2-afc0-a1948923e99f">